### PR TITLE
Allows files/folders to be copied.

### PR DESCRIPTION
**Pull request by tcitworld**
Adds a Copy action in the file action menu.
This also enables copy ~~and move~~ actions on selected files.

I know this is not how @jancborchardt designed it [here](https://github.com/nextcloud/server/issues/3206#issuecomment-297966461), and that the [files_clipboard](https://github.com/leizh/owncloud-files_clipboard) app is being developed but hey, this works for me, so it might be worth a look. :)
Ref. https://github.com/nextcloud/server/issues/3206

Needs : 
- [x] Tests
- [ ] ~~Copy action icon~~
- [ ] Tests on federated shares, remote storage, ...
- [ ] Tests for regressions on filepicker UI

The move feature has been moved at #6273

### CODE REVIEW

File: apps/files/js/filelist.js [hunk 1]
```diff
@@ -2000,10 +2072,94 @@
 						self.showFileBusyState($tr, false);
 					});
 			});
+			if (callback) {
+				callback();
```
Comment (tcitworld) [inline apps/files/js/filelist.js]: Dirty.

**Comment by MorrisJobke** [comment]
cc @danxuliu 


@tcitworld Could I ask you to split the "add copy" and the "move for selected files" into two PRs? Then we could work on them in parallel and one does not block the other feature.

**Comment by tcitworld** [comment]
Yup, will do that.

**Review by jancborchardt — DISMISSED** [review]
Nice work! However I would really like an addition like that to only come in with the design as specified at https://github.com/nextcloud/server/issues/3206#issuecomment-297966461

Could you adjust it to that? :)

**Comment by tcitworld** [comment]
Looks like this
selection_102

**Comment by jancborchardt** [comment]
@tcitworld nice! :) Was just mentioning it in the design lightning talk today. ;) Is the menu also changed to say »Move or Copy«? (Will test the whole flow later)

File: apps/files/js/fileactions.js [hunk 2]
```diff
@@ -618,30 +618,21 @@
 			});
 
 			this.registerAction({
-				name: 'Move',
-				displayName: t('files', 'Move'),
+				name: 'CopyMove',
+				displayName: t('files', 'Copy or Move'),
```
Comment (jancborchardt) [inline apps/files/js/fileactions.js]: Just a tiny one: First move, then copy. So »Move or copy« :) (with small _c_opy)

File: apps/files/templates/list.php [hunk 3]
```diff
@@ -47,9 +47,9 @@
 					</label>
 					<a class="name sort columntitle" data-sort="name"><span><?php p($l->t( 'Name' )); ?></span><span class="sort-indicator"></span></a>
 					<span id="selectedActionsList" class="selectedActions">
-						<a href="" class="copy">
+						<a href="" class="copy-move">
 							<span class="icon icon-external"></span>
-							<span><?php p($l->t('Copy'))?></span>
+							<span><?php p($l->t('Copy or Move'))?></span>
```
Comment (jancborchardt) [inline apps/files/templates/list.php]: Same here, »Move or copy« :)

File: core/js/oc-dialogs.js [hunk 4]
```diff
@@ -233,15 +247,47 @@ var OCdialogs = {
...
-				click: functionToCall,
-				defaultButton: true
-			}];
+
+			var chooseCallback = function () {
+				functionToCall(OCdialogs.FILEPICKER_TYPE_CHOOSE);
+			};
+
+			var copyCallback = function () {
+				console.log('copy callback');
```
Comment (icewind1991) [inline core/js/oc-dialogs.js]: leftover debug statement

File: apps/files/templates/list.php [hunk 5]
```diff
@@ -47,6 +47,10 @@
 					</label>
 					<a class="name sort columntitle" data-sort="name"><span><?php p($l->t( 'Name' )); ?></span><span class="sort-indicator"></span></a>
 					<span id="selectedActionsList" class="selectedActions">
+						<a href="" class="copy-move">
+							<span class="icon icon-external"></span>
+							<span><?php p($l->t('Copy or Move'))?></span>
```
Comment (icewind1991) [inline apps/files/templates/list.php]: Same thing here as @jancborchardt said

**Review by icewind1991 — DISMISSED** [review]
Looks good otherwise

**Comment by tcitworld** [comment]
Rebased now that #6273 is merged.

**Review by icewind1991 — APPROVED** [review]
works great 👍

**Comment by jancborchardt** [comment]
:+1: from my side. What about the failing tests @tcitworld @MorrisJobke?

**Comment by MorrisJobke** [comment]
> 👍 from my side. What about the failing tests @tcitworld @MorrisJobke?

Those are timeouts, that are solved in current master/stable12 and this is simply from an older version of them. In the future they should not happen at all (or at least not that often anymore).

**Comment by MorrisJobke** [comment]
> Tests
>  Tests on federated shares, remote storage, ...
>  Tests for regressions on filepicker UI

What about those pending items? @tcitworld Could you add at least some tests for this?

**Review by jancborchardt — APPROVED** [review]

**Comment by tcitworld** [comment]
I added tests and fixed a behaviour where failed copy actions were also counted as successful. Please review again.

**Comment by jancborchardt** [comment]
@tcitworld looks like the commits should be squashed into one? :) Also, there’s some Drone failures

**Comment by danxuliu** [comment]
> there’s some Drone failures

The failure was in an acceptance test... that should not have failed (I have run the test in local and it worked fine). Restarting the build will hopefully make the failure to go away (although I will have to study the failure in detail anyway; I hope it does not appear too frequently in Drone until then :-S ).

**Review by MorrisJobke — APPROVED** [review]

**Comment by agoat** [comment]
Nice work @tcitworld. Thanks.

**Comment by jancborchardt** [comment]
Yup, great work @tcitworld :)

**Pull request merged**